### PR TITLE
Refactoring of upgrade command

### PR DESCRIPTION
**Pull request by kosovvv**

### CODE REVIEW

File: Sitefinity CLI/Commands/UpgradeCommand.cs [hunk 1]
```diff
@@ -51,447 +45,97 @@ internal class UpgradeCommand
...
-            {
-                projectFilesAbsolutePaths = projectFilesAbsolutePaths
-                    .Where(ap => this.HasSitefinityReferences(ap) && this.HasValidSitefinityVersion(ap));
-            }
-
-            return projectFilesAbsolutePaths.ToList();
-        }
-
-        private bool HasValidSitefinityVersion(string projectFilePath)
+        private async Task Validate()
```
Comment (RuzmanovDev) [inline Sitefinity CLI/Commands/UpgradeCommand.cs:105]: There are several propblems in the Validate method. on Line 117 where we  if (!this.SkipPrompts && !this.promptService.PromptYesNo(Constants.UpgradeWarning)) we cancel the execution of the method with the return statement. However this will not cancel the exection of the command like it did before.
Also in the validate method we set the version property if it is not set. This is something odd. An method should do one thing and should not have such side effects

File: Sitefinity CLI/Services/ProjectService.cs [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+                {
+                    var configValue = File.ReadAllText(porjectConfigPath);
+                    configsWithoutSitefinity.Add(porjectConfigPath, configValue);
+                }
+            }
+
+            return configsWithoutSitefinity;
+        }
+
+        public IEnumerable<string> GetProjectPathsFromSolution(string solutionPath, string version, bool onlySitefinityProjects = false)
```
Comment (RuzmanovDev) [inline Sitefinity CLI/Services/ProjectService.cs:62]: The version property is used only when the optional parameter is true. If it is false it would not matter what is the value of the version property.
Maybe we can split it into two methods, One that retrieves all projects form the solution and one that retreives only sitefinity projects. The second one can have the version parameter. Also if there is a common logic between the methods we can reuse it (make a private method or something like that)

File: Sitefinity CLI/Commands/UpgradeCommand.cs [hunk 3]
```diff
@@ -51,447 +45,97 @@ internal class UpgradeCommand
...
 
-            Dictionary<string, string> configsWithoutSitefinity = this.GetConfigsForProjectsWithoutSitefinity(projectsWithouthSitefinityPaths);
-
             this.logger.LogInformation(string.Format(Constants.NumberOfProjectsWithSitefinityReferencesFoundSuccessMessage, sitefinityProjectFilePaths.Count()));
-            this.logger.LogInformation(string.Format("Collecting Sitefinity NuGet package tree for version \"{0}\"...", this.Version));
-            var packageSources = await this.packageSourceBuilder.GetNugetPackageSources(this.NugetConfigPath);
+            this.logger.LogInformation(string.Format(Constants.CollectionSitefinityPackageTreeMessage, this.Version));
 
-            NuGetPackage newSitefinityPackage = await this.sitefinityPackageManager.GetSitefinityPackageTree(this.Version, packageSources);
+            NuGetPackage installedPackage = await this.packageService.InstallPackage(UpgradeOptions, sitefinityProjectFilePaths);
```
Comment (RuzmanovDev) [inline Sitefinity CLI/Commands/UpgradeCommand.cs:92]: For better readability we can construct the parameter we pass to the Install package method here. Also it is strange the naming is odd. This method is doing more that just installing the package.

File: Sitefinity CLI/Services/PackageService.cs [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+                        compatiblePackage = package;
+                        break;
+                    }
+                }
+            }
+
+            return compatiblePackage;
+        }
+
+        public async Task<NuGetPackage> InstallPackage(UpgradeOptions options, IEnumerable<string> sitefinityProjectFilePaths)
```
Comment (RuzmanovDev) [inline Sitefinity CLI/Services/PackageService.cs:53]: The method is doing more than installing the package. Maybe we should rename and extract some logic.
I see that here we ultimately download the new nuget package in order to be able to build the config used for upggrading sitefinity. We can leave the restore, set target framework and install (this.sitefinityPackageManager.Install) here and name the method something like "PrepareSitefinityUpgradePackage". The output of the method will be a NugetPackage object which we can pass the the UgpradeConfigGenerator

File: Sitefinity CLI/Services/Interfaces/IPackageService.cs [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+using Sitefinity_CLI.Model;
+
+namespace Sitefinity_CLI.Services.Interfaces
+{
+    public interface IPackageService
+    {
+        Task<NuGetPackage> GetLatestCompatibleVersion(string packageId, Version sitefinityVersion, string nugetConfigPath);
+        Task<NuGetPackage> InstallPackage(UpgradeOptions options, IEnumerable<string> sitefinityProjectFilePaths);
+        void SyncProjectReferencesWithPackages(IEnumerable<string> projectFilePaths, string solutionFolder);
+
```
Comment (RuzmanovDev) [inline Sitefinity CLI/Services/Interfaces/IPackageService.cs:14]: Remove empty line and add one after each method declaration

File: Sitefinity CLI/Services/VisualStudioService.cs [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+                    this.logger.LogError(string.Format("Error occured while upgrading nuget packages. {0}", result));
+                    throw new UpgradeException("Upgrade failed");
+                }
+
+                break;
+            }
+
+            this.logger.LogInformation("Operation completed successfully!");
+        }
+        private string ReadAllTextFromFile(string path)
```
Comment (RuzmanovDev) [inline Sitefinity CLI/Services/VisualStudioService.cs:81]: Add empty lines between methods

File: Sitefinity CLI/Services/Interfaces/IProjectService.cs [hunk 7]
```diff
@@ -0,0 +1,16 @@
+﻿using Sitefinity_CLI.Model;
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Sitefinity_CLI.Services.Interfaces
+{
+    public interface IProjectService
+    {
+        Task<string> GetLatestSitefinityVersion();
```
Comment (RuzmanovDev) [inline Sitefinity CLI/Services/Interfaces/IProjectService.cs:10]: add empty line between method declarations

File: Sitefinity CLI/Services/VisualStudioService.cs [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+using Sitefinity_CLI.PackageManagement;
+using Sitefinity_CLI.Services.Interfaces;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+
+namespace Sitefinity_CLI.Services
+{
+    internal class VisualStudioService : IVisualStudioService
```
Comment (RuzmanovDev) [inline Sitefinity CLI/Services/VisualStudioService.cs:13]: We have VisualStudio worker that is encapsulating the VS related operations. (you use it here)
This class is not so related to visual studio. Also you have declared one method called InitializeSolution. 
In this method you are doing several stuffs. Maybe we should rename the class to VisualStuidoUpgradeService and the method should be. ExecuteVisualStudioUpgrade

File: Sitefinity CLI/Services/Interfaces/IPackageService.cs [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+using System.Collections.Generic;
+using Sitefinity_CLI.Model;
+
+namespace Sitefinity_CLI.Services.Interfaces
+{
+    public interface IPackageService
+    {
+        Task<NuGetPackage> GetLatestCompatibleVersion(string packageId, Version sitefinityVersion, string nugetConfigPath);
+        Task<NuGetPackage> InstallPackage(UpgradeOptions options, IEnumerable<string> sitefinityProjectFilePaths);
+        void SyncProjectReferencesWithPackages(IEnumerable<string> projectFilePaths, string solutionFolder);
```
Comment (RuzmanovDev) [inline Sitefinity CLI/Services/Interfaces/IPackageService.cs:13]: This is not related to NugetPackages. Should not be part of this service.

File: Sitefinity CLI/Services/ProjectService.cs [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+using System.Linq;
+using System.Net.Http;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using Sitefinity_CLI.Model;
+using Microsoft.Extensions.Logging;
+
+namespace Sitefinity_CLI.Services
+{
+    internal class ProjectService : IProjectService
```
Comment (RuzmanovDev) [inline Sitefinity CLI/Services/ProjectService.cs:18]: Project service is very generic naming. Also the methods that are included are not related to "Projects"
GetLatestSitefinityVersion()  GetConfigsForProjectsWithoutSitefinity() RestoreReferences,

Think of a better naming. Maybe extract different services or add the methods in some of the existing ones

File: Sitefinity CLI/Services/PackageService.cs [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+using Sitefinity_CLI.Services.Interfaces;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Sitefinity_CLI.Services
+{
+    internal class PackageService : IPackageService
```
Comment (RuzmanovDev) [inline Sitefinity CLI/Services/PackageService.cs:13]: Again like in the project service. Package Service is too generic

**Comment by RuzmanovDev** [comment]
General note. When extracting services think of what the service would do. What will be its responsibilities. What will be its dependencies. Make sure that the extracted services have single responsibility and methods that can be easily tested

File: Sitefinity CLI/PackageManagement/Contracts/ISitefinityPackageManager.cs [hunk 12]
```diff
@@ -19,10 +20,13 @@ internal interface ISitefinityPackageManager
 
         Task<NuGetPackage> GetPackageTree(string id, string version, IEnumerable<NugetPackageSource> nugetPackageSources, Func<NuGetPackage, bool> shouldBreakSearch = null);
 
-        Task<IEnumerable<string>> GetPackageVersions(string id, int versionsCount = 10);
+        Task<IEnumerable<string>> GetPackageVersions(string id, IEnumerable<NugetPackageSource> packageSources, int versionsCount = 10);
 
         void SyncReferencesWithPackages(string projectFilePath, string solutionFolder);
 
+        // TODO: not part  of this interface
```
Comment (RuzmanovDev) [inline Sitefinity CLI/PackageManagement/Contracts/ISitefinityPackageManager.cs:27]: Remove todo

**Review by RuzmanovDev — APPROVED** [review]

**Review by gpeshev1 — APPROVED** [review]

**Review by gpeshev1 — APPROVED** [review]
